tasks/actorstate/actorstate.go

```diff
@@ -255,8 +255,15 @@ func (p *ActorStateProcessor) processBatch(ctx context.Context, node lens.API) (
 		ctx, cancel = context.WithDeadline(ctx, claimUntil)
 		defer cancel()
 	}
-
 	stats.Record(ctx, metrics.TipsetHeight.M(batch[0].Height))
+
+	var doneBatch visor.ProcessingActorList
+	defer func() {
```

Comment:
There may be a race here. When using leasing we set up a context with a deadline based on the lease. When that expires the remainder of the batch is abandoned and we need to make sure that anything we completed is updated in the database. However the context has been cancelled so any database updates will fail. We could probably fix that here by reordering where we create the deadline context so it's not passed into the update.

However, I think this is a wider problem in general and suspect that some of the issues we see with bad connections and advisory locks not being dropped is due to the use of cancelled contexts in the database calls.